### PR TITLE
new read counting mode for v4, cd-hit-dup fix for unpaired reads

### PR DESCRIPTION
**Pull request by boris-dimitrov**
This code change behaves differently based on the pipeline major version.   See deployment plan below.

USER-VISIBLE CHANGES IN VERSION >= 4

  *  In the report, both "R" and "Contig R" will change to reflect ReadCountingMode.COUNT_ALL rather than ReadCountingMode.COUNT_UNIQUE.

  * Similarly, "rPM" will change to reflect the count of ALL reads for its corresponding taxon divided by the count of all non-ercc reads for the entire sample.  In other words, the new rRM values will be equal to the old values multiplied by the DCR (duplicate compression ratio) for their corresponding taxons.  That's better, because the old values were off by a factor of DCR, making comparisons less meaningful across projects.
 
USER-VISIBLE CHANGES IN VERSION 3.21

  *  The pipeline will emit a new cdhitdup_cluster_sizes.tsv file with one line per unique read, showing the cluster size for the read followed by the read id. 

  *  The pipeline will emit the raw clustering output of cdhit, where the user could see the set of all read ids for each cluster.

  *  The pipeline will emit taxon_count_with_dcr.json files which explicitly show the count of all reads, unique reads, and their ratio, which is called DCR.

  *  The behavior of cd-hit-dup will be correct even for unpaired reads.  Previously, for unpaired inputs, cd-hit-dup could emit only half of the unique reads, and add duplicates of those.

  *  The cd-hit-dup read identity threshold will be 100% rather than 95%, for the first 70 nucleotides.

DEPLOYMENT PLAN

(1a) Merge this to master.  That deploys it in the sense that all the new files will be emitted, but counts (all values visible on the report page) will remain unchanged. 

(1b) When ready to flip the switch, merge a one-liner PR that simply sets `idseq_dag.__version__` to 4.0.   This will automatically flip ReadCountingMode from COUNT_UNIQUE to COUNT_ALL.

POST-DEPLOYMENT

This can happen immediately after (1a):

(2)  Port to idseq-web the dag changes that this PR hackplements in pipeline_flow.

(3) When the idseq-web change is deployed, revert the pipeline_flow hack from here.

Steps 2 and 3 are not gating anything, they are code cleanliness steps that are invisible to users.

### CODE REVIEW

File: idseq_dag/util/m8.py [hunk 1]
```diff
@@ -508,13 +538,15 @@ def any_hits_to_remove(hits):
...
+                                'count': 0,  # todo:  rename to nonunique_count -- since count originally meant unique_count
+                                'unique_count': 0,
                                 'sum_percent_identity': 0.0,
                                 'sum_alignment_length': 0.0,
                                 'sum_e_value': 0.0
                             }
                             aggregation[agg_key] = agg_bucket
-                        agg_bucket['count'] += 1
+                        agg_bucket['count'] += get_read_dcr(cdhit_cluster_sizes, read_id)
+                        agg_bucket['unique_count'] += 1
```
Comment (boris-dimitrov) [inline idseq_dag/util/m8.py]: I missed a few more places to change below, where we average percent_identity the denominator has to be unique_count.

**Comment by boris-dimitrov** [comment]
Initial test results here.

This PR (pipeline 3.101):
https://staging.idseq.net/samples/14567

Old experiment that just simply disabled cdhitdup (pipeline 3.99):
https://idseq.net/samples/39223

Original prod behavior (pipeline 3.16):
https://idseq.net/samples/39223?pipelineVersion=3.16

**Comment by boris-dimitrov** [comment]
cc @MarkAZhang since this touches on some of the read-reassignment code.

**Comment by boris-dimitrov** [comment]
TLDR from review with Katrina and getting background info from Charles:

  (*)  Don't worry about MIN_CONTIG_SIZE, that filter will be deprecated soon

  (*)  A single per-taxon DCR should be emitted in a separate file, with no requirement to show this on the report.

File: idseq_dag/engine/pipeline_flow.py [hunk 2]
```diff
@@ -79,6 +79,22 @@ def parse_and_validate_conf(dag_json):
...
+                cdhitdup_cluster_sizes_path
+            ]
+            dag["given_targets"][cdhitdup_cluster_sizes_target] = {
+                "s3_dir": dag["given_targets"]["host_filter_out"]["s3_dir"]
+            }
+        if dag["name"] == "host_filter":
+            for fff in ["dedup1.fa.clstr", cdhitdup_cluster_sizes_path]:
+                if fff not in dag["targets"]["cdhitdup_out"]:
+                    dag["targets"]["cdhitdup_out"].append(fff)
+
```
Comment (gregdingle) [inline idseq_dag/engine/pipeline_flow.py]: for clarity, I'd like to see these lines extracted to a private method, and linked to a JIRA ticket that tracks the clean up work needed

File: idseq_dag/engine/pipeline_flow.py [hunk 3]
```diff
@@ -87,6 +103,10 @@ def parse_and_validate_conf(dag_json):
         dag['name'] = dag.get("name", _get_name_from_path(dag_json))
         covered_targets = set()
         for s in steps:
+            # hack -- augmnet steps as needed for testing of cdhitdup
+            if s["class"] in ("PipelineStepBlastContigs", "PipelineStepRunAlignmentRemotely", "PipelineStepRunAssembly"):
+                if cdhitdup_cluster_sizes_target not in s["in"]:
+                    s["in"].append(cdhitdup_cluster_sizes_target)
```
Comment (gregdingle) [inline idseq_dag/engine/pipeline_flow.py]: same comment as above

File: idseq_dag/steps/blast_contigs.py [hunk 4]
```diff
@@ -231,37 +227,68 @@ def run_blast(db_type, blast_m8, *args):
...
-            species_summary[species_taxid][contig] += 1
-            genus_summary[genus_taxid][contig] += 1
-        # constructing the array for output
+            record_read(species_taxid, genus_taxid, contig, read_id)
+
+        # Filter out contigs that contain too few unique reads.
+        # This used to happen in db_loader in idseq-web.  Any code left there that still appears to
+        # do this filtering is effectively a no-op and the filtering cannot be done there because
+        # the non-unique read counts are no longer output by the pipeline.
+        for summary in [species_summary, genus_summary]:
```
Comment (gregdingle) [inline idseq_dag/steps/blast_contigs.py]: extract to private method

File: idseq_dag/steps/blast_contigs.py [hunk 5]
```diff
@@ -231,37 +227,68 @@ def run_blast(db_type, blast_m8, *args):
...
 
         for read_id, read_info in added_reads_dict.items():
             contig = read2contig[read_id]
             species_taxid, genus_taxid, _family_taxid = contig2lineage[contig]
-            species_summary[species_taxid][contig] += 1
-            genus_summary[genus_taxid][contig] += 1
-        # constructing the array for output
+            record_read(species_taxid, genus_taxid, contig, read_id)
+
+        # Filter out contigs that contain too few unique reads.
```
Comment (gregdingle) [inline idseq_dag/steps/blast_contigs.py]: good comment!

File: idseq_dag/steps/blast_contigs.py [hunk 6]
```diff
@@ -231,37 +227,68 @@ def run_blast(db_type, blast_m8, *args):
...
+                for contig in list(contig_counts.keys()):
+                    unique_count, nonunique_count = contig_counts[contig]
+                    if unique_count < MIN_CONTIG_SIZE:
+                        del contig_counts[contig]
+                    else:
+                        contig_counts[contig] = nonunique_count
+                if not contig_counts:
+                    del summary[taxid]
+
+        # construct the array for output
```
Comment (gregdingle) [inline idseq_dag/steps/blast_contigs.py]: extract to private method for bonus points :)

File: idseq_dag/util/m8.py [hunk 7]
```diff
@@ -56,6 +60,10 @@
     }
 }
 
+# The minimum read count for a valid contig. We ignore contigs below this read count in most downstream analyses.
+# This constant is hardcoded in at least 4 places in idseq-web.  TODO: Make it a DAG parameter.
+MIN_CONTIG_SIZE = 4
```
Comment (gregdingle) [inline idseq_dag/util/m8.py]: good comment. please make a JIRA ticket for TODO.

File: idseq_dag/util/m8.py [hunk 8]
```diff
@@ -419,13 +427,35 @@ def call_hit_level_v2(hits):
                     msg += f"\t{species_taxid}\t{genus_taxid}\t{family_taxid}\n"
                     outf_sum.write(msg)
 
+
+def get_read_dcr(cdhit_cluster_sizes, read_id):
```
Comment (gregdingle) [inline idseq_dag/util/m8.py]: maybe rename to `get_read_cluster_size` ?

File: idseq_dag/steps/run_cdhitdup.py [hunk 9]
```diff
@@ -40,6 +49,45 @@ def run(self):
             )
         )
 
+        # Emit cluster sizes.  One line per cluster.  Cluster size followed by cluster read id.
+        # The rest of the pipeline operates on unique reads (i.e. on the representatives of these clusters).
+        # In the end, cluster sizes are used to convert unique read counts to original read counts.
+        with open(cdhit_cluster_sizes_path, "w") as tsv,\
```
Comment (gregdingle) [inline idseq_dag/steps/run_cdhitdup.py]: extract to method in util or other file. ideally write some unit tests.

File: idseq_dag/engine/pipeline_flow.py (hunk 2 above)
Comment (cdebourcy) [reply idseq_dag/engine/pipeline_flow.py]: I believe these targets need to be added to the DAG json in idseq-web, otherwise it's incompatible with the new SFN-WDL infra -- @kislyuk is that correct? Seems like the current code would require some ad-hoc migration

Comment (boris-dimitrov) [reply idseq_dag/engine/pipeline_flow.py]: Yes, first we merge this PR into idseq dag, then we deploy an idseq-web change to emit correct DAG, then we remove this hack from idseq-dag.

Comment (boris-dimitrov) [reply idseq_dag/engine/pipeline_flow.py]: I've updated the main PR description with the resolutions to all resolved questions, current list of todos, and most importantly, the deployment plan.

Comment (boris-dimitrov) [reply idseq_dag/engine/pipeline_flow.py]: The hack lines above are really not meant to live in master for more than 1-2 weeks tops.  Basically they can be reverted here as soon as we deploy the idseq-web change that does the same (with may be a couple extra days to make sure we won't have to revert that change).

File: idseq_dag/engine/pipeline_step.py [hunk 10]
```diff
@@ -262,3 +266,27 @@ def step_resources(self):
         By default, show link to idseq-dag documentation.
         '''
         return {"IDseq Docs": "https://github.com/chanzuckerberg/idseq-dag/wiki"}
+
+
+class PipelineCountingStep(PipelineStep):
```
Comment (gregdingle) [inline idseq_dag/engine/pipeline_step.py]: it would be nice to have some docblock here such as

```
PipelineStep that counts nonunique reads based on back-calculation from cluster sizes 
TSV file emitted by `PipelineStepRunCDHitDup`. Most steps should subclass this class. 
```

File: idseq_dag/engine/pipeline_step.py [hunk 11]
```diff
@@ -262,3 +266,27 @@ def step_resources(self):
...
+
+    def input_cluster_sizes(self):
```
Comment (gregdingle) [inline idseq_dag/engine/pipeline_step.py]: nit: I wouldn't expect such a function name to return a file path. How about `input_cluster_sizes_path` ?

File: idseq_dag/util/count.py [hunk 12]
```diff
@@ -70,7 +61,7 @@ def lines2reads(line_count, file_format):
         read_count = line_count
     return read_count
 
-def reads2lines(read_count, file_format):
+def _reads2lines(read_count, file_format):
```
Comment (gregdingle) [inline idseq_dag/util/count.py]: I like the `_` to denote private 👍

File: idseq_dag/util/count.py [hunk 13]
```diff
@@ -88,7 +79,100 @@ def files_have_min_reads(input_files, min_reads):
...
+def reads(local_file_path):
+    '''
+    Count reads in given FASTA or FASTQ file.
+    Implemented via wc, so very fast.
+    '''
+    return _count_reads_via_wc(local_file_path, max_reads=None)
+
+def reads_in_group(file_group, max_fragments=None, cluster_sizes=None):
+    '''
+    OVERVIEW
```
Comment (gregdingle) [inline idseq_dag/util/count.py]: awesome documentation! 🌟

File: idseq_dag/util/count.py [hunk 14]
```diff
@@ -88,7 +79,100 @@ def files_have_min_reads(input_files, min_reads):
...
+    3-4 million reads per minute. Fortunately, only steps run_lzw and run_bowtie2 can
+    have more than a million fragments, and even that is extremely unlikely (deeply
+    sequenced microbiome samples are rare in idseq at this time).  All other steps either
+    operate on at most 1 million fragments or do not specify cluster_sizes. If this
+    performance ever becomes a concern, the relevant private function can be reimplemented
+    easily in GO.
+
+    INTERFACE
+
+    It may be better to expose the two use cases as separate functions, so that users do
```
Comment (gregdingle) [inline idseq_dag/util/count.py]: Lols, I don't think the interface is that important here because calls to this function will be added very rarely from now on.

File: idseq_dag/util/count.py [hunk 15]
```diff
@@ -88,7 +79,100 @@ def files_have_min_reads(input_files, min_reads):
...
+       (*) Count FASTA or FASTQ fragments without expanding clusters,
+           optionally truncating to max_fragments.  The input may
+           even be gz compressed.
+
+    PERFORMANCE
+
+    When cluster_sizes is not specified, the implementation is very fast, via wc.
+
+    When cluster_sizes is specified, the python implementation can process only about
+    3-4 million reads per minute. Fortunately, only steps run_lzw and run_bowtie2 can
```
Comment (gregdingle) [inline idseq_dag/util/count.py]: good to see you benchmarked. how much total runtime do you expect this to add to a typical run?

File: idseq_dag/engine/pipeline_step.py (hunk 10 above)
Comment (boris-dimitrov) [reply idseq_dag/engine/pipeline_step.py]: great, thanks

File: idseq_dag/engine/pipeline_step.py (hunk 11 above)
Comment (boris-dimitrov) [reply idseq_dag/engine/pipeline_step.py]: sounds good

File: idseq_dag/util/count.py (hunk 15 above)
Comment (boris-dimitrov) [reply idseq_dag/util/count.py]: for most runs, less than 15 seconds total

File: idseq_dag/util/count.py (hunk 14 above)
Comment (boris-dimitrov) [reply idseq_dag/util/count.py]: Fortunately, yes.   In case someone tries to use it in future, I felt it good to warn them.

**Comment by boris-dimitrov** [comment]
After correct counts and per-taxon DCR:   https://staging.idseq.net/samples/14747   (in progress)

**Comment by boris-dimitrov** [comment]
Some correctness testing.

For the same input data, compare pipeline versions

   3.16  <--->  3.20  (counting unique reads):
   https://idseq.net/samples/39223?pipelineVersion=3.16
   https://staging.idseq.net/samples/14758?pipelineVersion=3.19   (it's really 3.20 never mind it says 3.19)

   3.99 <---> 4.101  (counting all reads):
   https://idseq.net/samples/39223?pipelineVersion=3.99
   https://staging.idseq.net/samples/14918?pipelineVersion=4.101

Minor differences are due to an index update.

The version that will be deployed with this PR is 3.20;  flipping major version to 4.0 will unlock the results from 4.101 above.

File: idseq_dag/engine/pipeline_flow.py [hunk 16]
```diff
@@ -79,6 +80,25 @@ def parse_and_validate_conf(dag_json):
           "given_targets": input files that are given
         '''
         dag = json.loads(open(dag_json).read())
+
+        # hack -- add some targets here to support cd-hit-dup work
+        if DAG_SURGERY_HACKS_FOR_READ_COUNTING:
```
Comment (gregdingle) [inline idseq_dag/engine/pipeline_flow.py]: 👍
